### PR TITLE
[fix/docs]: improve the Jump Game algorithm

### PR DESCRIPTION
**Pull request by Panquesito7**
#### Description of Change

- Improve the Jump Game algorithm.
  - Fix segmentation fault.
  - Improved documentation.
  - Added missing namespace (`greedy_algorithms`).
  - Used proper filename with snake_case.
  - Added 3 extra tests.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by Panquesito7** [comment]
I just noticed something very wrong in my algorithm 😅 works in the tests but not for everything. Will fix it ASAP.

File: greedy_algorithms/jump_game.cpp [hunk 1]
```diff
@@ -67,11 +69,11 @@ static void test() {
     assert(greedy_algorithms::can_jump(nums) == true);
 
     // 4th test
-    nums = { 4, 2, 8, 9, 6 };
+    nums = { 1, 0, 5, 8, 12 };
     assert(greedy_algorithms::can_jump(nums) == false);
```
Comment (CaedenPH) [inline greedy_algorithms/jump_game.cpp]: Why not just keep the original tests and add new ones?

File: greedy_algorithms/jump_game.cpp [hunk 2]
```diff
@@ -40,12 +40,14 @@ namespace greedy_algorithms {
  * @returns false if the index can NOT be reached
  */
 template <typename T>
-bool can_jump(const std::vector<T> &nums, const int &index = 1) {
-    const int size = nums.size() + 1 - nums[index];
-    if (nums[index] >= size) { // `>=` because the number can be higher than the size of the array.
-        return true;
+bool can_jump(const std::vector<T> &nums) {
+    size_t lastPos = nums.size() - 1;
+    for (int i = nums.size() - 1; i >= 0; i--) {
```
Comment (CaedenPH) [inline greedy_algorithms/jump_game.cpp:45]:
```suggestion
    for (int i = lastPos; i >= 0; i--) {
```

File: greedy_algorithms/jump_game.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+namespace greedy_algorithms {
+/**
+ * @brief Checks whether the given element (default is `1`) can jump to the last
+ * index.
+ * @param nums array of numbers containing the maximum jump (in steps) from that
+ * index
+ * @returns true  if the index can be reached
+ * @returns false if the index can NOT be reached
+ */
+template <typename T>
```
Comment (vil02) [inline greedy_algorithms/jump_game.cpp:42]: Is the `T` really needed here? In your problem description you write about _an array of non-negative integers_. What do you think about simply having `std::size_t`?

**Review by vil02 — CHANGES_REQUESTED** [review]

File: greedy_algorithms/jump_game.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+bool can_jump(const std::vector<T> &nums) {
+    size_t lastPos = nums.size() - 1;
```
Comment (vil02) [inline greedy_algorithms/jump_game.cpp:44]:
```suggestion
    const size_t lastPos = nums.size() - 1;
```

File: greedy_algorithms/jump_game.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    nums = { 5, 9, 4, 7, 15, 3 };
+    assert(greedy_algorithms::can_jump(nums) == true);
+
+    // 4th test
+    nums = { 1, 0, 5, 8, 12 };
+    assert(greedy_algorithms::can_jump(nums) == false);
+
+    // 5th test
+    nums = {2, 1, 4, 7};
+    assert(greedy_algorithms::can_jump(nums) == true);
```
[the remark this answers is not in the document]
Comment (realstealthninja) [reply greedy_algorithms/jump_game.cpp:77]: Your suggestion looks really clean!

File: greedy_algorithms/jump_game.cpp (hunk 4 above)
Comment (Panquesito7) [reply greedy_algorithms/jump_game.cpp:44]: Not possible, as it is being modified on the loop.

File: greedy_algorithms/jump_game.cpp (hunk 3 above)
Comment (Panquesito7) [reply greedy_algorithms/jump_game.cpp:42]: I'm going to stick with `int` for now. Nothing else can be used at all.

File: greedy_algorithms/jump_game.cpp (hunk 4 above)
Comment (vil02) [reply greedy_algorithms/jump_game.cpp:43]: You are right. I have overlooked that. Sorry for the confusion.

File: greedy_algorithms/jump_game.cpp (hunk 3 above)
Comment (vil02) [reply greedy_algorithms/jump_game.cpp]: > Nothing else can be used at all.

What do you mean?

I would highly recommend not to use `int` here. If you want to use `int`, you should throw an exception of one of the entries of `nums` is negative.

**Review by realstealthninja — APPROVED** [review]

**Pull request merged**